src/zope/keyreference/persistent.py

```diff
@@ -96,8 +96,16 @@ def _get_cmp_keys(self, other):
         return self.key_type_id, other.key_type_id
 
     # Py3: For Python 2 BBB.
+    # If we implement all the rich comparison operations, though, this is
+    # never actually called.
     def __cmp__(self, other):
-        return cmp(*self._get_cmp_keys(other))
+        my_keys, other_keys = self._get_cmp_keys(other)
+        if my_keys == other_keys:
+            return 0
+        if my_keys > other_keys:
+            return 1
+        return -1
```

Comment:
Do we need to have it at all?  "Python 2 BBB" is questionable to me, given that Python 2 will use rich comparison in preference to `__cmp__`.

Reply:
Given the interaction of proxies and `__cmp__` as outlined in https://github.com/zopefoundation/zope.keyreference/pull/6, it wasn't clear to me that it was safe to drop it.

Reply:
I did some experimentation and confirmed that even under Python 2.7, `__cmp__` isn't used at all, even when proxies are involved, regardless of using `==` or `cmp()` directly. One would have to be calling `__cmp__()` explicitly, which one should generally not be doing. 

Of course, `__cmp__` is documented in `IKeyReference`, so there might be callers or subclasses (which are probably pointless or broken).

So I think we can remove `__cmp__` from the implementations and the interfaces. (We'd still have the zope.security issue, which needs to be fixed there, not here.) As for whether that removal requires a major version bump...strictly speaking, probably yes, but OTOH they've been effectively pointless since 4.0a1.

Thoughts?

Reply:
I've no strong feelings either way.

(Maybe merge this, release, then drop the dead code and bump version?)